freedomofpress/securedrop-proxy
# Improve error handling, tests

## Description

Try to ensure the proxy always returns a valid JSON response. Add specific handling of upstream error responses to proxy.py, and add tests and fixtures for that.

Replace the Proxy._on_done static method with an instance method.

In main.py, bail out upon receiving invalid input JSON, instead of printing an error response and then continuing to try to use it, resulting in printing another error response.

Have main.py use the callbacks on the proxy as given, instead of always overwriting them with the defaults from callbacks.py.

Rework entrypoint.py so that it should always print a JSON response, and add tests for it.

Fixes #13.

## Testing

1. Build the `securedrop-proxy` source distribution:
    - check out this branch
    - run `python setup.py sdist`
2. Build the `.deb`:
    - in a clone of `securedrop-debian-packaging`, run `PKG_PATH=/home/user/src/securedrop-proxy/dist/securedrop-proxy-0.1.6.tar.gz PKG_VERSION=0.1.6 make securedrop-proxy` (adjust the path to your working copy of this repo as necessary)
3. Install the `.deb` in `sd-svs-buster-template`:
    - `qvm-copy ~/debbuild/packaging/securedrop-proxy_0.1.6+buster_all.deb` and specify `sd-svs-buster-template` as the target VM
    - In a shell on `sd-svs-buster-template`, run:
        - `cp ~/QubesIncoming/sd-dev/securedrop-proxy_0.1.6+buster_all.deb /tmp`
        - `sudo apt install /tmp/securedrop-proxy_0.1.6+buster_all.deb`
4. Reboot.
5. Start the SecureDrop client in `sd-svs` and confirm that the proxy is communicating with the server: submissions and replies should populate with no errors.
